port-chooser/src/main/java/org/terracotta/utilities/test/net/PortManager.java

```diff
@@ -0,0 +1,568 @@
+   * really be free.  According to issue <a href="https://jira.terracotta.org/browse/MNK-5621">MNK-5621</a>,
+   * some Windows services, like Remote Desktop (RDP) don't establish an open listener
+   * but do respond to connection requests on their assigned port(s).  A "failure to connect"
+   * is necessary to determine if the port is actually available.  (This check presumes
+   * firewall rules aren't responsible for dropping the connection request -- not much we
+   * can do about that.)
+   * @param port the port to test
+   * @return {@code true} if the port is available (free); {@code false} otherwise
+   * @see <a href="https://support.microsoft.com/en-us/help/832017/service-overview-and-network-port-requirements-for-windows">
+   *   Service overview and network port requirements for Windows</a>
+   */
+  private boolean refusesConnect(int port) {
+    InetSocketAddress endpoint;
+    if (LOCALHOST == null) {
+      endpoint = new InetSocketAddress("localhost", port);
+    } else {
+      endpoint = new InetSocketAddress(LOCALHOST, port);
+    }
+
+    boolean isFree;
+    Socket sock = new Socket();
+    //noinspection TryFinallyCanBeTryWithResources
+    try {
+      sock.connect(endpoint, 50);
+      isFree = false;     // Connected: not free
+    } catch (SocketTimeoutException e) {
+      isFree = true;      // Connection attempt timed out; **free**
+    } catch (Exception e) {
+      LOGGER.debug("[refusesConnect] Client connection to port {} failed for unexpected reason", endpoint, e);
+      isFree = false;     // Other error; can't assess -- presumed not free
+    }  finally {
+      try {
+        sock.close();
+      } catch (IOException ignored) {
+      }
+    }
+
+    return isFree;
+  }
+
+
+  /**
+   * {@code Reference} implementation used to track no longer used reserved ports.
+   */
+  private static class AllocatedPort extends WeakReference<PortRef> {
+    private final int port;
+    private final AtomicReference<Runnable> closer;
+    private AllocatedPort(PortRef referent, ReferenceQueue<? super PortRef> q) {
+      super(referent, q);
+      this.port = referent.port();
+      this.closer = referent.closers();
+    }
+
+    private void release() {
+      this.closer.get().run();
+    }
+  }
+
+  /**
+   * Represents a reserved TCP port.  When use of the port is no longer needed, this
+   * reference should be closed to release the reservation.
+   */
+  public static class PortRef implements AutoCloseable {
+    private final int port;
+    private final AtomicReference<Runnable> closers = new AtomicReference<>(() -> {});
+
+    private PortRef(int port) {
+      this.port = port;
+    }
+
+    /**
+     * Prepends a close action to this {@code PortRef}.  Actions are executed
+     * in reverse order of their addition.
+     * @param action the action to take to close this {@code PortRef} instance
+     */
+    void onClose(Runnable action) {
+      closers.accumulateAndGet(action, PortManager::combine);
```

Comment:
This method behaves oddly after the ref is closed.

Reply:
Addressed